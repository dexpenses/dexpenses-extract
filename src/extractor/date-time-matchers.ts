import { DateTime } from 'luxon';
import MatcherDef, {
  withSanityCheck,
  statically,
  withReplacements,
} from '../utils/matcher/matcher-def';

function currentYear() {
  return new Date().getFullYear();
}

export default {
  d: /(?:[1-9]|[12]\d|3[01])/,
  dd: /(?:0[1-9]|[12]\d|3[01])/,
  M: /(?:[1-9]|1[0-2])/,
  MM: /(?:0[1-9]|1[0-2])/,
  MMM: /(?:jan|feb|mÃ¤rz|apr|mai|jun|jul|aug|sep|okt|nov|dez)/i,
  yyyy: withSanityCheck(
    /(?:19|2\d)\d{2}/,
    (m) => parseInt(m, 10) <= currentYear()
  ),
  yy: withSanityCheck(
    /[1-6][0-9]/,
    (m) =>
      DateTime.fromFormat(m, 'yy', {
        zone: 'Europe/Berlin',
      }).year <= currentYear()
  ),
  h: withReplacements(/(?:[1i][0-2i]|[1-9])/i, ['i', 1]),
  HH: withReplacements(/(?:[01i][0-9i]|2[0-4i])/i, ['i', 1]),
  mm: withReplacements(/[0-5i][0-9i]/, ['i', 1]),
  ss: withReplacements(/[0-5i][0-9i]/, ['i', 1]),
  ':': statically(/ ?[:;] ?/),
  a: /[AP]M/i,
  '^': statically(/(?:^| )/, ''),
  $: statically(/(?: |$)/, ''),
  '.': statically(/ ?[\.,] ?/),
  '-': statically(/ ?- ?/),
} as Record<string, MatcherDef>;
