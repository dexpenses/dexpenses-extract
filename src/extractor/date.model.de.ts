export default [
  'dd.MM.yy HH:mm',
  'dd.MM.yyyy',
  'dd.MM.yy',
  'dd-MM-yyyy',
  'd-M-yyyy',
  'dd/MM/yyyy',
  'ddMM-yyyy',
  'ddMMyyyy',
  'd.MM.yyyy',
  'd.MM.yy',
];
