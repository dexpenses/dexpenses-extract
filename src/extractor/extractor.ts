import { Receipt } from '@dexpenses/core';

export type Opt<T> = T | undefined | null;

export abstract class Extractor<T> {
  constructor(public readonly field: string) {}

  public abstract extract(
    text: string,
    lines: string[],
    extracted: Receipt
  ): Opt<T> | Promise<Opt<T>>;
}
