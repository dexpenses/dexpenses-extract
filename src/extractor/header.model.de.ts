import HeaderExtractorOptions from './header.model';

export default {
  maxHeaderLines: 8,
  minLineLength: 2,
  headerDelimiters: [
    { pattern: /[\d\w]/, negative: true },
    /^\s*Artikelname\s*$/i,
    /^\s*Preis:?\s*$/i,
    /^UID\sNr/i,
    /^\s*EUR\s*$/i,
    /^\s*\d+[,.]\d\d\s*$/i,
    /^\s*St\.?Nr\.?/i,
    /[oö]ffnungszeit(en)?/i,
    /(^|\s)Kartenzahlu[np]g($|\s)/i,
    /(^|\s)Bezahlung($|\s)/i,
    /^€/i,
    /^\d+\sCashier$/i,
    { pattern: /(Quittung|Rechnung)/i, minLine: 5 },
    /mo-(fr|sa|so):? ?([01]\d|2[0-4])( ?:? ?[0-5]\d)? ?(-|bis) ?([01]\d|2[0-4])( ?:? ?[0-5]\d)? ?(uhr)?/i,
  ],
  irrelevantPatterns: [
    /(^| )[*XжN ]*(Karten|K[uv]nden|H(ä|ae?)ndler)bele[gqa][*XжN]*( |$)/gim,
    /(^| ) ?-? ?K ?-? ?[a-z0-9\- ]+B ?-? ?E ?-? ?L ?-? ?E ?-? ?[Gqa] ?-? ?/gim, // Kundenbeleg with " ?-? ?" between letter matchers
    /Bedient von: [a-z]+/i,
    /www?\s?\.\s?[a-z\-]+\s?\.\s?[a-z]+(\/[a-z\-_\d%]+)*\/?/i,
    /Vielen Dank\s?(,[a-z\s&äüö]+[\?!.])?/gim,
    /Bis bald!?/i,
    /Obj(\.|ekt)-?Nr\.?:?\s?\d+(\s|$)/i,
    /K[SA]\.\s?\d+/i,
    /(Term(inal)?[\- ]?ID|TA[- ]Nr\.?) ?:?( ?\d+)?/gim,
    /(^| )ID ?:? ?\d+/gim,
    /^:\s*/,
    /Bed\.?Nr\.?:?\s?\d+/i,
    /Kasse\s?\d+/i,
    /BNr\.?\s?\d+/i,
    /wir haben f(ü|ue?)r sie ge(ö|oe?)ffnet[:!\.]?/i,
    /t(ä|ae?)glich ab \d+( uhr)?/i,
    /rgnr\.?:?\s?\w+/i,
    /Zahlungsart/i,
    /Kellner:?\s?\d+/i,
    /Tisch:?\s?\d+/i,
    /(^| )Arbeitszettel( |$)/gim,
    /(sagt )?danke f(ü|ue?)r Ihren Einkauf/i,
    /(und )?auf Wiedersehen/i,
    /Bon[\- ]?ID:?\s?(\d+)?/im,
    /wir danken\sf(ü|ue?)r ihren einkauf[!.]?/gim,
    /[HW](ä|ae?)hrung:? ?(EURO?)?/gim,
    /(^| )Rechnung( |$)/gim,
    / ?-? ?Q ?-? ?U ?-? ?I ?-? ?T ?-? ?T ?-? ?U ?-? ?N ?-? ?G ?-? ?/gim,
    /Bedienung:? ?/,
    /(Steuer|USt)[ \-]?(ID)?(Nr)?\.?:? ?(DE ?)?[\d\/]+/i,
    /^Lieferzeit$/gim,
    /Tankstellen-?Nr\.?:? ?\d+/gi,
    /^\d{1,4}$/gm,
    /(^| )(Beleg|Trace)(\s*\-?\s*Nr\.?|nummer):? ?(\d+)?/gim,
    /(^| )Datum([ :]|$)/gim,
    /(^| )[UJ]hrzeit:?([ a-z\d():;]*Uhr)?( |$)/gim,
    /^Uhr$/gim,

    /(^| )Zwischensumme( |$)/gim,
    /(?<=\n)\d+ ?total ?\d*( |$)/gim,

    /(^| )Fax[.:]?( ?[\dog\-\/ ]+)?( |$)/gim,
    /(^| )\(?\s?[O0]rtstarif\s?\)?( |$)/gim,

    /^[a-z][^a-z\d]$/gim, // indicate wrongly detected text
  ],
  trimPatterns: [
    /^[ *жxN]+$/i,
    /^[ *xжN]*[ *ж]/i,
    /[ *ж][ *жxN]*$/i,
    /^[,\/]/,
    /[,\/]$/,
  ],
  fixes: [
    {
      pattern: /(^|\s)(?:6mbH|GabH|Gmbh)(\s|$)/g,
      replaceWith: '$1GmbH$2',
    },
    {
      pattern: /^(.*[a-z].*)GMBH(\s|$)/g,
      replaceWith: '$1GmbH$2',
    },
  ],
} as HeaderExtractorOptions;
