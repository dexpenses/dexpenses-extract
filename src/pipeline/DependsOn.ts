import Stage from './Stage';

type StageCtor = new (...args: any[]) => Stage;

export function DependsOn(...value: StageCtor[]) {
  return (target: StageCtor) => {
    if (value.indexOf(target) !== -1) {
      throw new Error(`${target.name} cannot depend on itself`);
    }
    target.prototype.$dependsOn = value;
  };
}

export function checkDependencies(stages: Array<Stage | Stage[]>) {
  for (const [i, stage] of stages.entries()) {
    for (const subStage of Array.isArray(stage) ? stage : [stage]) {
      if (!subStage.$dependsOn) {
        continue;
      }
      const precedingStages = stages.slice(0, i).flat();
      for (const dependency of subStage.$dependsOn) {
        if (!precedingStages.find((s) => s.constructor === dependency)) {
          throw new Error(
            `Dependency '${dependency.name}' not satisfied on ${subStage.constructor.name}`
          );
        }
      }
    }
  }
}
