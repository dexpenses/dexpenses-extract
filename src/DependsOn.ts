import { Extractor } from './extractor/extractor';

type Class = new (...args: any[]) => {};

export function DependsOn(...value: Class[]) {
  return function MyClassDecorator(target: any): any {
    if (value.indexOf(target) !== -1) {
      throw new Error(`${target.name} cannot depend on itself`);
    }
    target.prototype.$dependsOn = value;
  };
}

export function checkDependencies(pipeline: Array<Extractor<any>>) {
  pipeline.forEach((extractor, index) => {
    const dependencies = (extractor as any).$dependsOn;
    if (!dependencies) {
      return;
    }
    for (const dependency of dependencies) {
      if (!pipeline.slice(0, index).find((e) => e instanceof dependency)) {
        throw new Error(
          `Dependency '${dependency.name}' not satisfied on ${
            extractor.constructor.name
          }`
        );
      }
    }
  });
}
