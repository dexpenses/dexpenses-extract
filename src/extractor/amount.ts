import { Extractor } from './extractor';
import { Receipt, Amount } from '@dexpenses/core';
import { DependsOn } from '../DependsOn';
import { PaymentMethodExtractor } from './paymentMethod';
import {
  getAllMatches,
  anyMatches as anyRegexMatches,
} from '../utils/regex-utils';
import { anyMatches } from './util';

@DependsOn(PaymentMethodExtractor)
export class AmountExtractor extends Extractor<Amount> {
  constructor() {
    super('amount');
  }

  public extract(
    text: string,
    lines: string[],
    extracted: Receipt
  ): Amount | null {
    const amount = anyMatches(text, [
      /(?:gesamt|summe)(?:\s+EUR)?\s*(\d+,\d\d).*$/i,
      /betrag(?:\s+EUR)?\s*(\d+,\d\d).*$/i,
      // /^geg(?:\.|eben)(?:\sVISA)?$(?:\s+EUR)?\s*(\d+,\d\d).*$/im,
      /^geg(?:\.|eben)\sVISA$(?:\s+EUR)?\s*(\d+,\d\d).*$/im,
      /^(\d+,\d\d)$\n^Total in EUR$/im,
      /(\d+,\d\d)(?:$\n^eur)?$\n^zu zahlen/im,
      /total:?(?:$\n)?^(\d+,\d\d)$/im,
    ]).then(
      (m) =>
        ({
          value: parseFloat(m[1].replace(',', '.')),
          currency: 'EUR',
        } as Amount)
    );
    if (amount) {
      return amount;
    }
    if (extracted.paymentMethod === 'CASH') {
      const amountValue = findAmountFromCashPaymentValues(
        getAmountValues(lines)
      );
      if (amountValue) {
        return {
          value: amountValue,
          currency: 'EUR',
        };
      }
    }
    const maxAmount = getAmountValues(lines).reduce(
      (max: number | null, cur) => {
        return !max || cur > max ? cur : max;
      },
      null
    );
    if (maxAmount) {
      return {
        value: maxAmount,
        currency: 'EUR',
      };
    }
    return null;
  }
}

/*
 1. line start or space
 2. optionally negative number (, and . match as decimal point with optional space after decimal point)
 with no leading zero where S also matches as 5
 3. line end, space or dash
 */
const amountValuePattern = /(?:^|\s|\*)-?((?:[1-9]\d+|\d)\s?[,.]\s?[\dS]{2})(?:[\-\sâ‚¬]|$)/gim;

const illegalPreviousLinePatterns = [/MwSt\.?\s?%?$/i];

const illegalAmountPrefixPatterns = [
  /AS-Zeit:?\s?$/i,
  /punktestand entspricht:?\s?$/i,
  /MwSt:?\s?$/i,
];

const illegalAmountSuffixPatterns = [/^\s?%/, /^\s?Uhr/i];

// TODO: we could include date filter (i.e. not take dd.MM from the matched date as amount value)
export function getAmountValues(lines: string[]): number[] {
  return lines
    .flatMap((line, lineIndex) =>
      getAllMatches(amountValuePattern, line).map((match) => ({
        match,
        lineIndex,
      }))
    )
    .filter(
      ({ match, lineIndex }) =>
        (lineIndex === 0 ||
          !anyRegexMatches(
            lines[lineIndex - 1],
            illegalPreviousLinePatterns
          )) &&
        !anyRegexMatches(
          match.input.slice(0, match.index),
          illegalAmountPrefixPatterns
        ) &&
        !anyRegexMatches(
          match.input.slice(match.index + match[0].length),
          illegalAmountSuffixPatterns
        )
    )
    .map(({ match: [_, amount] }) => looselyParseFloat(amount));
}

function looselyParseFloat(s: string): number {
  return parseFloat(
    s
      .replace(/\s/g, '')
      .replace(/S/g, '5')
      .replace(',', '.')
  );
}

function equal(x: number, y: number) {
  return Number(Math.abs(x - y).toFixed(2)) < Number.EPSILON;
}

export function findAmountFromCashPaymentValues(values: number[]) {
  for (let i = values.length - 3; i >= 0; i -= 1) {
    const [amount, given, back] = values.slice(i, i + 3);
    if (equal(amount + (back < 0 ? -back : back), given)) {
      return amount;
    }
  }
  return null;
}
