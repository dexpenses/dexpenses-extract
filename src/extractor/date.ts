import { DateTime } from 'luxon';
import model from './date.model.de';
import { DependsOn } from '../DependsOn';
import { Extractor } from './extractor';
import { cleanHeaders, HeaderExtractor } from './header';
import { Receipt } from '@dexpenses/core';
import {
  createMatcher,
  Matcher,
  RegExpMatcher,
  statically,
} from './util/matcher';

export const matchers: Record<string, RegExpMatcher> = {
  d: /([1-9]|[12]\d|3[01])/,
  dd: /(0[1-9]|[12]\d|3[01])/,
  M: /([1-9]|1[0-2])/,
  MM: /(0[1-9]|1[0-2])/,
  yyyy: /([12]\d{3})/,
  yy: /([1-6][0-9])/,
  HH: /([01]\d|2[0-4])/,
  mm: /([0-5]\d)/,
  ss: /([0-5]\d)/,
  '.': statically(/\s?[\.,]\s?/),
  '-': statically(/\s?\-\s?/),
};

@DependsOn(HeaderExtractor)
export class DateExtractor extends Extractor<Date> {
  private matcher: Matcher;

  constructor() {
    super('date');
    this.matcher = createMatcher(matchers, model);
  }

  public extract(text: string, lines: string[], extracted: Receipt) {
    return this.matcher.exec(text).then((res) => {
      const fullDate = res.polishedMatch();

      cleanHeaders(extracted, fullDate);
      return DateTime.fromFormat(fullDate, res.def.format, {
        zone: 'Europe/Berlin',
      })
        .set({ hour: 0, minute: 0, second: 0 })
        .toJSDate();
    });
  }
}
