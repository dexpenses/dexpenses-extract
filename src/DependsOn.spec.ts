import { checkDependencies, DependsOn } from './DependsOn';
import { Extractor } from './extractor/extractor';
import { Receipt } from '@dexpenses/core';

class ExtractorA extends Extractor<any> {
  public extract(text: string, lines: string[], extracted: Receipt) {
    throw new Error('Method not implemented.');
  }
}
@DependsOn(ExtractorA)
class ExtractorB extends Extractor<any> {
  public extract(text: string, lines: string[], extracted: Receipt) {
    throw new Error('Method not implemented.');
  }
}
@DependsOn(ExtractorA, ExtractorB)
class ExtractorC extends Extractor<any> {
  public extract(text: string, lines: string[], extracted: Receipt) {
    throw new Error('Method not implemented.');
  }
}

describe('DependsOn', () => {
  it('should fail if dependency of an extractor is placed after the extractor in the pipeline', () => {
    expect(() =>
      checkDependencies([new ExtractorB(''), new ExtractorA('')] as any)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should fail if second dependency of an extractor is placed after the extractor in the pipeline', () => {
    expect(() =>
      checkDependencies([
        new ExtractorA(''),
        new ExtractorC(''),
        new ExtractorB(''),
      ] as any)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should successfully verify dependencies if correct order is provided (1 dependency)', () => {
    expect(() =>
      checkDependencies([new ExtractorA(''), new ExtractorB('')])
    ).not.toThrowError();
  });

  it('should successfully verify dependencies if correct order is provided (2 dependencies)', () => {
    expect(() =>
      checkDependencies([
        new ExtractorA(''),
        new ExtractorB(''),
        new ExtractorC(''),
      ])
    ).not.toThrowError();
  });
});
