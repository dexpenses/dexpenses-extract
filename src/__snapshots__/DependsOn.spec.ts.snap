// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependsOn should fail if dependency of an extractor is placed after the extractor in the pipeline 1`] = `"Dependency 'ExtractorA' not satisfied on ExtractorB"`;

exports[`DependsOn should fail if second dependency of an extractor is placed after the extractor in the pipeline 1`] = `"Dependency 'ExtractorB' not satisfied on ExtractorC"`;
